src/compiler/crystal/semantic/cleanup_transformer.cr

```diff
@@ -551,27 +557,29 @@ module Crystal
         then_is_truthy = false
         then_is_falsey = false
       else
-        node.then = node.then.transform(self)
-        then_is_truthy, then_is_falsey = @last_is_truthy, @last_is_falsey
+        then_is_truthy, then_is_falsey = compute_last_truhiness do
+          node.then = node.then.transform(self)
+        end
       end
 
       if node.truthy?
         else_is_truthy = false
         else_is_falsey = false
       else
-        node.else = node.else.transform(self)
-        else_is_truthy, else_is_falsey = @last_is_truthy, @last_is_falsey
+        else_is_truthy, else_is_falsey = compute_last_truhiness do
+          node.else = node.else.transform(self)
+        end
       end
 
-      reset_last_status
-
       case node
       when .and?
         @last_is_truthy = cond_is_truthy && then_is_truthy
         @last_is_falsey = cond_is_falsey || then_is_falsey
       when .or?
-        @last_is_truthy = (cond_is_truthy && then_is_truthy) || (cond_is_falsey && else_is_truthy)
+        @last_is_truthy = cond_is_truthy || else_is_truthy
```

Comment:
The bug was this line, which computes whether an `or` expression like `a || b` is always truthy or falsey.

As documented, something like:

```crystal
a || b
```

is turned into:

```crystal
if temp = a # cond
  temp # then
else
  b # else
end
```

So seeing the above we can say that `a || b` is truthy when either `a` or `b` are truthy. But since the `||` is turned into the `if` above, it's when the cond part is truthy or when the else part is truthy (for the cond part the compiler will see through an assignment to check whether the assignment value is truthy).

Also note the symmetry of the resulting code with `and`.

The bug was that the old condition was nonsense: I didn't fully understand it at that time and just experimented with some combinations... shame on me. Or actually, I thought the version in this PR was correct but it didn't work for some reason. And the reason was that the variables `@last_is_truthy` and `@last_is_falsey` weren't reset before reaching this line, so their value was wrong.

So this PR fixes those things.

Reply:
Oh, and all of this is computed in order to know whether we can ignore some parts of an if: if the condition is truthy then there's no need to codegen the `else` part, we can always codegen the `then` part. Same but reversed when the condition is `falsey`.